hq_superset/views.py

```diff
@@ -99,34 +105,30 @@ class CCHQApiException(Exception):
 
 
 def refresh_hq_datasource(domain, datasource_id, display_name):
-    # This method pulls the data from CommCareHQ and creates/replaces the
-    #   corresponding Superset dataset
-    datasource_url = get_datasource_export_url(domain, datasource_id)
+    """
+    Pulls the data from CommCare HQ and creates/replaces the
+    corresponding Superset dataset
+    """
     provider = superset.appbuilder.sm.oauth_remotes["commcare"]
-    oauth_token = get_valid_cchq_oauth_token()
-    response = provider.get(datasource_url, token=oauth_token)
-    if response.status_code != 200:
-        # Todo; logging
-        raise CCHQApiException("Error downloading the UCR export from HQ")
-    zipfile = ZipFile(BytesIO(response.content))
-    filename = zipfile.namelist()[0]
-    # Upload to table
+    token = get_valid_cchq_oauth_token()
     database = get_ucr_database()
     schema = get_schema_name_for_domain(domain)
     csv_table = Table(table=datasource_id, schema=schema)
-
+    datasource_defn = get_datasource_defn(provider, token, domain, datasource_id)
+    column_dtypes, date_columns = get_column_dtypes(datasource_defn)
     try:
-        df = pd.concat(
-            pd.read_csv(
-                chunksize=1000,
-                filepath_or_buffer=zipfile.open(filename),
-                encoding="utf-8",
-                # Todo; make date parsing work
-                parse_dates=True,
-                infer_datetime_format=True,
-                keep_default_na=True,
+        with get_csv_file(provider, token, domain, datasource_id) as csv_file:
```

Comment:
Nitpicking here (feel free to ignore), the context process needs to be just for the ZipFile.

Reply:
You are right that the context processor is for `ZipFile`. What I was thinking was that I could move more code out of the `refresh_hq_datasource()` function, and using a context process closes both the `ZipFile`, and the CSV file inside it. My primary objective was to move the code for fetching the CSV file into its own function. The context processor was just convenient.